### DISCUSSION TITLE
Maintainers bots and actions

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#15019] [Ideas]
I've been trying to think of ways to increase our number of packages with official maintainers. Here's what I came up with.

1. Add a GitHub action to automatically add package `maintainers` as reviewers of PRs. This was attempted in #12269, but later removed because we never got it working. I wanted to open a dedicated issue on this so we remember to try again.
2. Add a GitHub action to automatically add package `maintainers` as assignees for issues. This is a little harder, as we can't directly see which package is modified, it needs to be inferred from the issue title. I propose we standardize on `package: ` or `[package]` in the issue title.
3. Add a bot with a standardized "Would you like to add yourself as a maintainer? Here's how. Here's why." message that comments on all PRs for packages without a maintainer. This could get a little annoying, but hopefully it will convince people to add themselves as maintainers.

**Comment by adamjstewart** [comment]
This looks like a useful tutorial: https://github-bot-tutorial.readthedocs.io/en/latest/index.html

**Comment by adamjstewart** [comment]
These also look interesting:
https://github.com/marketplace/actions/github-action-for-assignee-to-reviewer
https://github.com/marketplace/actions/auto-assign-action

**Comment by tgamblin** [comment]
I'm pretty positive we need a bot.  I've sort of given up on actions doing more than CI -- at the moment you can't do much that is trusted inside a github action, so I think we will end up having to use bots for this stuff.  That's fine, I suppose, I just wish it were a bit simpler to run this kind of task in an action.

Example: adding a PR reviewer [requires that the reviewer has write access to the repo](https://developer.github.com/v3/pulls/review_requests/).  So we would need to add maintainers to a team and give them write access (note that we would still, as we do now, protect release branches, `develop`, and `master`).  And, the requesting account has to be at least a collaborator with read access.  You can't currently do that in an action, especially for PRs from forks, because the action [doesn't run with sufficient privileges](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token).

A bot, on the other hand, can have whatever token we give it.  But we need to host it somewhere.  We can do that in AWS -- I will just need to set it up.

My suggestion for this would be to start working on bot(s) that can run in containers, and make repos for them in the Spack org.  Then we can start throwing them into our AWS kube cluster and running them.  The containers will need to have a way to inject a token (e.g., through an env var).  That should give us what we need.

**Comment by vvolkl** [comment]
Another suggestion:

- parsing the `tags` of a package and adding them to PRs that change them as labels.

I also thought of using a github actions workflow, but the limitations with the permissions for forks are really frustrating :(

**Comment by adamjstewart** [comment]
That should be pretty doable. I believe we're currently working on getting a Kube cluster setup at LLNL, so once that happens I'll start working on that.

**Comment by vvolkl** [comment]
@adamjstewart @tgamblin great! actually it seems like github has adressed the issue with the actions workflow since your last comment:

https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target

It sounds like `pull_request_target` could be used to implement a pull request labeller. But I'm not sure it will work for the maintainers issue, so if it can be implemented with a bot, all the better.
